File: BSvsZP-Common -Java/Common/AgentList.java
package Common;

import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Iterator;
//import java.util.concurrent.locks.Lock;

import java.util.logging.Logger;

import org.omg.CORBA.portable.ApplicationException;

public class AgentList extends DistributableObject implements Iterable<AgentInfo> 
{
	private static final Logger log = Logger.getLogger(AgentList.class.getName());
	private ArrayList<AgentInfo> agents = new ArrayList<>();
    private Object myLock = new Object();
    private static int MinimumEncodingLength;
    private Lock lock = new Lock();

    public int FindIndex(short id)
    {
    	int result = -1;
    	int i = 0;
    	while (i < agents.size())
    	{
    		if (agents.get(i).getId() == id)
    			result =i;
    	}
    	return result; 
    }
    public short getClassId() {
        return (short) DISTRIBUTABLE_CLASS_IDS.AgentList.getValue();
    }

    public static int getMinimumEncodingLength() {
        MinimumEncodingLength = 4 // Object header
        						+ 2;           // Components
        return MinimumEncodingLength;
    }

    public AgentList() {
    }

    public static AgentList Create(ByteList bytes) throws ApplicationException, Exception {
        AgentList result = new AgentList();
        result.Decode(bytes);
        return result;
    }

    public void Dispose() {
        Clear();
    }

    public int Count() {
        int result = 0;
        synchronized (myLock) {
            result = agents.size();
        }
        return result;

    }

    public AgentInfo getAgentInfo(int index) {
        AgentInfo result = null;
        if (index >= 0 && index < agents.size()) {
            result = agents.get(index);
        }
        return result;
    }

    public void Add(AgentInfo agentInfo) {
        synchronized (myLock) {
            agents.add(agentInfo);
        }
    }

    public void Remove(AgentInfo agentInfo) {
        if (agentInfo != null) {
            Remove(agentInfo.getId());
        }
    }

    public void Remove(short agentId) {
        synchronized (myLock) {
            for (AgentInfo agent : agents) {
                if (agent.getId() == agentId) {
                    agents.remove(agent);
                    break;
                }
            }
        }
    }

    public void Clear() {
        synchronized (myLock) {
            agents.clear();
        }

    }

    @Override
    public Iterator<AgentInfo> iterator() {
        return agents.iterator();
    }

    @Override
    public void Encode(ByteList bytes) throws UnknownHostException, Exception {
        bytes.Add((short) DISTRIBUTABLE_CLASS_IDS.AgentList.getValue());                             // Write out the class type
        bytes.update();
        short lengthPos = bytes.getCurrentWritePosition();   // Get the current write position, so we
        bytes.update();                                  // can write the length here later

        bytes.Add((short) 0);                           // Write out a place holder for the length
        bytes.update();
        lock.lock();
        {
            bytes.Add((short) (agents.size()));
            bytes.update();
            for (AgentInfo component : agents) {
                bytes.Add(component);
                bytes.update();
            }
        }
        lock.unlock();
        short length = (short) (bytes.getCurrentWritePosition() - lengthPos - 2);
        bytes.WriteInt16To(lengthPos, length);          // Write out the length of this object    

    }

    @Override
    protected void Decode(ByteList bytes) throws ApplicationException, Exception {
        if (bytes == null || bytes.getRemainingToRead() < getMinimumEncodingLength()) {
            throw new ApplicationException("Invalid byte array", null);
        } else if (bytes.PeekInt16() != (short) DISTRIBUTABLE_CLASS_IDS.AgentList.getValue()) {
            throw new ApplicationException("Invalid class id", null);
        } else {
            short objType = bytes.GetInt16();

            short objLength = bytes.GetInt16();

            bytes.SetNewReadLimit(objLength);
            bytes.update();
            synchronized (myLock) {
                Clear();

                short count = bytes.GetInt16();

                for (int i = 0; i < count; i++) {
                    agents.add((AgentInfo) bytes.GetDistributableObject());
                }
            }

            bytes.RestorePreviosReadLimit();
        }

    }
    
    public void Update(AgentList agentList)
    {
        if (agentList!=null && agentList.Count()>0)
        	synchronized (myLock)
        	{
        		for (AgentInfo updatedAgent: agentList)
        		{
	                int index;
	                if ((index = FindIndex(updatedAgent.getId())) != -1)
	                	agents.set(index, updatedAgent);
	                else
	                    agents.add(updatedAgent);
	                
            }
        }
    }
    
    public AgentInfo FindClosestToLocation(FieldLocation location, ArrayList<AgentInfo.PossibleAgentType[]> types)
    {
        double closestDistance = 0;
        AgentInfo closestAgent = null;

        synchronized (myLock)
        {
            for(AgentInfo agent : agents)
            {
                double distance = FieldLocation.Distance(location, agent.getLocation());
                if ((types==null || types.size() ==0 || types.contains(agent.getAgentType()))
                        && (closestAgent == null || distance < closestDistance))
                {
                    closestAgent = agent;
                    closestDistance = distance;
                }
            }
        }

        return closestAgent;
    }

}

File: BSvsZP-Common/Common/AgentList.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using log4net;

namespace Common
{
    public class AgentList : DistributableObject, IEnumerable<AgentInfo>
    {
        #region Private Data Members
        private static readonly ILog log = LogManager.GetLogger(typeof(AgentList));

        private static Int16 ClassId { get { return (Int16)DISTRIBUTABLE_CLASS_IDS.AgentList; } }
        private List<AgentInfo> agents = new List<AgentInfo>();
        private object myLock = new object();
        #endregion

        #region Public Properties and Other Stuff
        public static int MinimumEncodingLength
        {
            get
            {
                return 4              // Object header
                       + 2;           // Components
            }
        }

        #endregion
      
        #region Constructors, Factories, and Destructors
        public AgentList() { }

        /// <summary>
        /// Factor method to create an object of this class from a byte list
        /// </summary>
        /// <param name="bytes">A byte list from which the distributable object will be decoded</param>
        /// <returns>A new object of this class</returns>
        new public static AgentList Create(ByteList bytes)
        {
            AgentList result = new AgentList();
            result.Decode(bytes);
            return result;
        }

        public void Dispose()
        {
            Clear();
        }

        #endregion

        #region Public Methods and Properties
        public object Lock { get { return myLock; } }
        public int Count
        {
            get
            {
                int result = 0;
                lock (myLock) { result = agents.Count; }
                return result;
            }
        }

        public int FindIndex(Int16 id)
        {
            int result = -1;
            result = agents.FindIndex(delegate(AgentInfo agent) { return agent.Id == id; });
            return result;
        }

        public AgentInfo this[int index]
        {
            get
            {
                AgentInfo result = null;
                if (index >= 0 && index < agents.Count)
                    result = agents[index];
                return result;
            }
        }

        public void Add(AgentInfo agentInfo)
        {
            lock (myLock)
            {
                agents.Add(agentInfo);
            }
        }

        public void Remove(AgentInfo agentInfo)
        {
            if (agentInfo != null)
                Remove(agentInfo.Id);
        }

        public void Remove(Int16 agentId)
        {
            lock (myLock)
            {
                foreach (AgentInfo agent in agents)
                    if (agent.Id == agentId)
                    {
                        agents.Remove(agent);
                        break;
                    }
            }
        }


        public void Update(AgentList agentList)
        {
            log.Debug("Enter AgentList Update");
            if (agentList!=null && agentList.Count>0)
            lock (myLock)
            {
                foreach (AgentInfo updatedAgent in agentList)
                    UpdateAgent(updatedAgent);
            }
        }

        public void UpdateAgent(AgentInfo updatedAgent)
        {
            int index;
            if ((index = FindIndex(updatedAgent.Id)) != -1)
            {
                if (updatedAgent.AgentStatus == AgentInfo.PossibleAgentStatus.LostGame ||
                    updatedAgent.Strength <= 0)
                {
                    log.DebugFormat("Remove agent Id={0}, index={1}", updatedAgent.Id, index);
                    agents.RemoveAt(index);
                }
                else
                {
                    log.DebugFormat("Update agent Id={0}, index={1}", updatedAgent.Id, index);
                    agents[index] = updatedAgent;
                }
            }
            else
            {
                log.DebugFormat("Add agent Id={0}", updatedAgent.Id);
                agents.Add(updatedAgent);
            }
        }

        public AgentInfo FindClosestToLocation(FieldLocation location, params AgentInfo.PossibleAgentType[] types)
        {
            log.Debug("Enter FindClosestToLocation");
            double closestDistance = 0;
            AgentInfo closestAgent = null;

            lock (myLock)
            {
                foreach (AgentInfo agent in agents)
                {
                    log.DebugFormat("Consider, Id={0}, Type={1}, Strength={2}", agent.Id, agent.AgentType, agent.AgentStatus);
                    if ((types==null || types.Length==0 || types.Contains(agent.AgentType)) && agent.Strength>0)
                    {
                        double distance = FieldLocation.Distance(location, agent.Location);
                        log.DebugFormat("distance={0}", distance);
                        if (closestAgent == null || distance < closestDistance)
                        {
                            log.Debug("new closests");
                            closestAgent = agent;
                            closestDistance = distance;
                        }
                    }
                }
            }

            return closestAgent;
        }

        public void Clear()
        {
            lock (myLock)
            {
                agents.Clear();
            }
        }

        #endregion

        #region Encoding and Decoding methods

        /// <summary>
        /// This method encodes an object of this class into a byte list
        /// </summary>
        /// <param name="bytes"></param>
        public override void Encode(ByteList bytes)
        {
            bytes.Add(ClassId);                             // Write out the class type

            Int16 lengthPos = bytes.CurrentWritePosition;   // Get the current write position, so we
                                                            // can write the length here later

            bytes.Add((Int16) 0);                           // Write out a place holder for the length

            lock (myLock)
            {
                bytes.Add(Convert.ToInt16(agents.Count));
                foreach (AgentInfo component in agents)
                    bytes.Add(component);
            }
            Int16 length = Convert.ToInt16(bytes.CurrentWritePosition - lengthPos - 2);
            bytes.WriteInt16To(lengthPos, length);          // Write out the length of this object        
        }

        /// <summary>
        /// This method decodes a message from a byte list.  It can onlt be called from within the class hierarchy.
        /// </summary>
        /// <param name="messageBytes"></param>
        protected override void Decode(ByteList bytes)
        {
            if (bytes == null || bytes.RemainingToRead < MinimumEncodingLength)
                throw new ApplicationException("Invalid byte array");
            else if (bytes.PeekInt16() != ClassId)
                throw new ApplicationException("Invalid class id");
            else
            {
                Int16 objType = bytes.GetInt16();
                Int16 objLength = bytes.GetInt16();

                bytes.SetNewReadLimit(objLength);

                lock (myLock)
                {
                    Clear();
                    Int16 count = bytes.GetInt16();
                    for (int i = 0; i < count; i++)
                        agents.Add(bytes.GetDistributableObject() as AgentInfo);
                }

                bytes.RestorePreviosReadLimit();
            }
        }

        #endregion

        #region IEmunerator Interface
        public IEnumerator<AgentInfo> GetEnumerator()
        {
            return agents.GetEnumerator();
        }

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
        #endregion

    }
}
